lib/connections/service.go

```diff
@@ -439,7 +460,24 @@ func (s *service) dialDevices(ctx context.Context, now time.Time, cfg config.Con
 		}
 
 		dialTargets := s.resolveDialTargets(ctx, now, cfg, deviceCfg, nextDialAt, initial, priorityCutoff)
-		if conn, ok := s.dialParallel(ctx, deviceCfg.DeviceID, dialTargets); ok {
+		if len(dialTargets) > 0 {
+			queue = append(queue, dialQueueEntry{deviceCfg.DeviceID, dialTargets})
+		}
+	}
+
+	// Shuffle the connection queue, so that if we only try a limited set of
+	// devices (or they in turn have limits and we're trying to load balance
+	// over several) it won't be the same ones in the same order every time.
+	rand.Shuffle(queue)
+
+	// Cut the queue down to the limit, if any.
+	if connectionLimit > 0 && len(queue) > allowAdditional {
+		l.Debugf("Dialing %d devices out of %d because we've reached the connection limit (%d)", allowAdditional, len(queue), connectionLimit)
+		queue = queue[:allowAdditional]
+	}
+
+	for _, entry := range queue {
+		if conn, ok := s.dialParallel(ctx, entry.id, entry.targets); ok {
```

Comment:
Dials may fail, thus instead of limiting the number ahead of dialing I'd count dialed devices, and stop once the limit is reached.